renter/proto/formcontract.go

```diff
@@ -23,13 +25,17 @@ const (
 
 // FormContract forms a contract with a host. The resulting contract will have
 // renterPayout coins in the renter output.
-func FormContract(w Wallet, tpool TransactionPool, key ed25519.PrivateKey, host hostdb.ScannedHost, renterPayout types.Currency, startHeight, endHeight types.BlockHeight) (ContractRevision, []types.Transaction, error) {
+func FormContract(w Wallet, tpool TransactionPool, key ed25519.PrivateKey, host hostdb.ScannedHost, renterPayout types.Currency, startHeight, endHeight types.BlockHeight) (_ ContractRevision, _ []types.Transaction, err error) {
 	s, err := NewUnlockedSession(host.NetAddress, host.PublicKey, 0)
 	if err != nil {
 		return ContractRevision{}, nil, err
 	}
 	s.host = host
-	defer s.Close()
+	defer func() {
+		if e := s.Close(); e != nil {
+			err = multierror.Append(err, e)
+		}
+	}()
```

Comment:
If this fails and the connection keeps alive, the following lock requests will fail, won't it?

Reply:
This session only exists within the `FormContract` function, so no.

Reply:
I might not understand it correctly but even if `s.Close` fails and the current session is alive, can we call `NewSession` without "contract is locked by another party" error?

Reply:
Ah, I see. For `FormContract` the answer is still no, because the contract doesn't exist yet. But for `RenewContract` and other RPCs...maybe? If closing the session fails because the host already closed the connection, then the contract should be unlocked the next time you call `Lock`. If something weird happens to the network, then the connection could be left in a half-open state, which would cause the host to wait for the connection to time out (during which you'd be unable to acquire the lock). But I don't think failing to close the connection will be correlated with the host being stuck half-open. Usually a half-open connection only happens if the machine loses power or someone pulls the ethernet cable out.

My suggestion is that you add some internal logging for the errors you're worried about, and observe whether any of them occur in production. I'm not keen on cluttering my code without clear benefit.

Reply:
I think even if there is a half-open socket, the host can accept another connection and then get conflict. If I'm not mistaken, a half-open connection could happen if a switch gets overflowed. 

Btw, don't hosts change the default keepalive?